lib/active_model/serializer/adapter/json_sideload.rb

```diff
@@ -0,0 +1,185 @@
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonSideload < Json
+
+        # the list of what has already been serialized will be kept here to
+        # help avoid infinite recursion
+        #
+        # this should be a list of association_name to a list of objects
+        attr_accessor :associations
+
+        # When we are sideloading associations, we can more easily track what
+        # has been serialized, so that we avoid infinite
+        # recursion / serialization.
+        def initialize(*args)
+          @associations = {}
+
+          super(*args)
+        end
+
+        def serializable_hash(options = nil)
+          options ||= {}
+
+          serialize_hash(options)
+
+          singularize_lone_associations
+
+          @associations
+        end
```

Comment:
I was using `ActiveModel::Serializer.config.sideload_associations = true` in my other branch.

I realized, the downside to this approach is that it will recursively build every relationship (and a relationship's relationships) from a serializer - so -- I don't know if there is a way we want to specify against that? or.. maybe that's not a bad thing? idk. I'd need to use this in a real app to see how it behaves.